src/include/mpi.h.in

```diff
@@ -793,6 +793,11 @@ typedef void (MPI_Comm_errhandler_function)(MPI_Comm *, int *, ...);
 typedef void (MPI_File_errhandler_function)(MPI_File *, int *, ...);
 typedef void (MPI_Win_errhandler_function)(MPI_Win *, int *, ...);
 typedef void (MPI_Session_errhandler_function)(MPI_Session *, int *, ...);
+/* the _x callbacks accepts scalar inputs and an additional "void *" extra_state */
+typedef void (MPIX_Comm_errhandler_function_x)(MPI_Comm, int, void *, ...);
```

Comment:
Do we really need the varargs `...` interface here? By the time we have an extra context, whatever is needed can be passed over in the context. Moreover, from the ABI point of view, C vararg functions are a bit fuzzy.

Reply:
Good point. We probably don't need `...`. MPICH (likely OMPI as well) never implemented it anyway.

Reply:
Which is not a surprise at all,  as varargs interfaces are inherently non portable (unless all implementations agree) thus defeating the API portability of MPI.

Reply:
Updated without `...`.

Reply:
`typedef void MPI_Comm_errhandler_function(MPI_Comm *comm, int *error_code, . . . );` is mentioned in the MPI standard.
>The remaining arguments are “varargs” arguments whose number and meaning is implementation-dependent. An implementation should clearly document these arguments.

`varargs` are supported by the SysV x86 ABI (https://www.uclibc.org/docs/psABI-x86_64.pdf) and it seems not to be an ABI problem, as long as the user knows the signature of the callback (https://langdev.stackexchange.com/questions/2177/do-variadic-functions-need-to-have-a-different-calling-convention-to-regular-fun).

Reply:
Jeff, I'm very well aware that the standard defines callbacks with `vararg` interfaces. My claim is that these extra arguments, if any, are implementation defined and defeat API portability of user codes. In other words, the varargs are like current `MPIX_` extensions. In short, varags defeat/hinder portability of codes, and also hinder our ABI efforts, therefore we should NOT use them in new MPI features.

Reply:
OMPI does provide additional information in the `...`, allowing the callee to trace the root cause of the trigger for the error handler. I don't see any reason to remove it from the ABI.

Second, there is a claim above that Fortran callbacks do not work in OMPI. Is there any proof of that or it is just an uninformed statement ?

Reply:
> I don't see any reason to remove it from the ABI.

Please note that we are talking about a NEW callback interface, therefore we are not removing anything, we are free to design the new thing however we see fit. 

That being said, my point about avoiding `...` is to not encourage implementation-dependent interfaces/protocols. However, this position comes from my inner fascist dwarf attempting to lecture others about how to do things. That's simply not fine.

I retract my original position. If the user want to mess with the varargs that an implementation may or may not pass, it is up to the user to do so and under the user responsibility how to deal with ABI incompatibilities if switching MPI implementations at runtime.

@hzhou @jeffhammond The fact that Open MPI uses these varags interface is good enough of a precedent to support it. It is ultimately up to the user implementing the error handler callback whether to inspect or not these varargs under the documented rules of the MPI implementation. An implementation-agnostic callback would simply ignore the varargs and not look into them at all to ensure compile-time and runtime-compatibility across implementations.

> Second, there is a claim above that Fortran callbacks do not work in OMPI

They do work, or course, but the implementation does not take advantage of pointer arguments like `MPI_Comm*`, simply because the various `MPI_XXX_c2f/f2c` routines are not identity functions in Open MPI. Moreover, as MPI allows for C `int` different from Fortran `integer` (via `MPI_Fint`), the truth is that all C callbacks with pointer arguments for MPI handles are a bit of a nonsense, handles should always be passed by value.